### PR TITLE
Account activations

### PR DESCRIPTION
**Pull request by thoth-ky**
- Add mail activation for development environment

### CODE REVIEW

File: app/controllers/account_activations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class AccountActivationsController < ApplicationController
+  def edit
```
Comment (chalchuck) [inline app/controllers/account_activations_controller.rb:3]: There's soo much happening here...

Split it up to small methods that do a single task.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -5,8 +5,14 @@ def new
   def create
     user = User.find_by(email: params[:session][:email].downcase)
```
Comment (chalchuck) [inline app/controllers/sessions_controller.rb]: Single responsible methods, this method does 4 actions, please split it up to make more simple and perform only two responsibilities.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,7 +1,16 @@
 class UsersController < ApplicationController
+  before_action :get_user, only: %i[edit update show destroy]
+  before_action :logged_in_user, only: %i[index edit update]
+  before_action :correct_user, only: %i[edit update]
+  before_action :admin_user, only: :destroy
+  
+  def index
+    @users = User.where(activated: true).paginate(page: params[:page])
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: we never do ActiveRecord queries from the controller (User.where(activated: true), instead, create a scope on the model and use in the controller.

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -11,18 +20,59 @@ def new
   def create
     @user = User.new(user_params)
     if @user.save
-      log_in(@user)
-      flash[:success] = "Welcome to the Sample App!"
-      redirect_to @user
+      @user.send_activation_email
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: extract this piece of code to be run as a Callback on the User model.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -11,18 +20,59 @@ def new
...
+      flash[:info] = "Please check your email to activate your account."
+      redirect_to root_url
     else
-      flash[:warning] = "Something went wrong!"
       render 'new'
     end
   end
 
+  def edit
+    @user
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: What does this one do?

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -11,18 +20,59 @@ def new
...
+  def update
+    if @user.update_attributes(user_params)
+      redirect_to @user, success:"User details successfully updated."
+    else
+      render 'edit', error: "Update failed"
+    end
+  end
+
+  def destroy
+    @user.destroy
```
Comment (chalchuck) [inline app/controllers/users_controller.rb:41]: Add a condition to ensure that a user can not delete him/herself.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -11,18 +20,59 @@ def new
...
+      :password,
+      :password_confirmation
+    )
+  end
+
+  def get_user
+    @user = User.find(params[:id])
+  end
+
+  def logged_in_user
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: if this method is used in more than the user controller, please make it a helper function.

File: app/models/user.rb [hunk 8]
```diff
@@ -1,11 +1,49 @@
...
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
+  end
+
+  def activate
+    update_columns(activated: true, activated_at: Time.zone.now)
+  end
+
+  def send_activation_email
+    UserMailer.account_activation(self).deliver_now
```
Comment (chalchuck) [inline app/models/user.rb:50]: always when working with mailers, user deliver_later

```suggestion
    UserMailer.account_activation(self).deliver_later
```

this queues the email sending and when resources are available, the mail is sent asynchronously.

File: app/models/user.rb [hunk 9]
```diff
@@ -1,11 +1,49 @@
...
+  end
+
+  # Will returns true if the given a token matching the digest.
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
+  end
+
+  def activate
```
Comment (chalchuck) [inline app/models/user.rb]: when defining methods that write to the database we use a (BANG)! operator.

So you can rename the method to `activate!`

File: app/mailers/user_mailer.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class UserMailer < ApplicationMailer
+  def account_activation(user)
+    @user = user
+    mail to: user.email, subject: "Account Activation"
+  end
+
+  def password_reset
+    @greeting = "Hi"
+    mail to: "to@example.org", subject: "Reset Password"
```
Comment (chalchuck) [inline app/mailers/user_mailer.rb]: the to address is incorrect.

Comment (chalchuck) [reply app/mailers/user_mailer.rb]: where is the password reset LINK?

**Review by chalchuck — CHANGES_REQUESTED** [review]
Review and revert.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (thoth-ky) [reply app/controllers/users_controller.rb]: This renders the current user, the callback ensures we have extracted user form params[:id] for edit, show and update

File: app/models/user.rb (hunk 8 above)
Comment (thoth-ky) [reply app/models/user.rb:50]: ok. Thanks###

File: app/mailers/user_mailer.rb (hunk 10 above)
Comment (thoth-ky) [reply app/mailers/user_mailer.rb]: I was hoping to handle this when I do the password reset chapter

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (thoth-ky) [reply app/controllers/users_controller.rb:41]: done

File: app/controllers/users_controller.rb (hunk 7 above)
Comment (thoth-ky) [reply app/controllers/users_controller.rb]: done

**Pull request merged**